flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTriggeringTest.java

```diff
@@ -283,31 +300,494 @@ public void testStopPeriodicScheduler() throws Exception {
+		// Periodic
+		final CompletableFuture<CompletedCheckpoint> onCompletionPromise = coord.triggerCheckpoint(
+			System.currentTimeMillis(),
+			CheckpointProperties.forCheckpoint(CheckpointRetentionPolicy.NEVER_RETAIN_AFTER_TERMINATION),
+			null,
+			true,
+			false);
+
+		// checkpoint trigger will not finish since master hook checkpoint is not finished yet
+		manuallyTriggeredScheduledExecutor.triggerAll();
+		assertTrue(coord.isTriggering());
+
+		// trigger cancellation
+		manuallyTriggeredScheduledExecutor.triggerNonPeriodicScheduledTasks();
+		assertTrue(coord.isTriggering());
+
+		try {
+			onCompletionPromise.get();
+			fail("Should not reach here");
+		} catch (ExecutionException e) {
+			final Optional<CheckpointException> checkpointExceptionOptional =
+				ExceptionUtils.findThrowable(e, CheckpointException.class);
+			assertTrue(checkpointExceptionOptional.isPresent());
+			assertEquals(CheckpointFailureReason.CHECKPOINT_EXPIRED,
+				checkpointExceptionOptional.get().getCheckpointFailureReason());
+		}
+
+		// continue triggering
+		masterHookCheckpointFuture.complete("finish master hook");
+
+		manuallyTriggeredScheduledExecutor.triggerAll();
+		assertFalse(coord.isTriggering());
+		// it doesn't really trigger task manager to do checkpoint
+		assertEquals(0, taskManagerCheckpointTriggeredTimes.get());
+		assertEquals(0, coord.getTriggerRequestQueue().size());
+	}
+
+	@Test
+	public void testTriggerCheckpointInitializationFailed() throws Exception {
+		// create some mock Execution vertices that receive the checkpoint trigger messages
+		final ExecutionAttemptID attemptID1 = new ExecutionAttemptID();
+		ExecutionVertex vertex1 = mockExecutionVertex(attemptID1);
+
+		// set up the coordinator and validate the initial state
+		CheckpointCoordinatorConfiguration chkConfig = new CheckpointCoordinatorConfiguration(
+			600000,
+			600000,
+			0,
+			Integer.MAX_VALUE,
+			CheckpointRetentionPolicy.NEVER_RETAIN_AFTER_TERMINATION,
+			true,
+			false,
+			0);
+		CheckpointCoordinator coord = new CheckpointCoordinator(
+			new JobID(),
+			chkConfig,
+			new ExecutionVertex[] { vertex1 },
+			new ExecutionVertex[] { vertex1 },
+			new ExecutionVertex[] { vertex1 },
+			new TestingCheckpointIDCounter(),
+			new StandaloneCompletedCheckpointStore(1),
+			new MemoryStateBackend(),
+			Executors.directExecutor(),
+			manuallyTriggeredScheduledExecutor,
+			SharedStateRegistry.DEFAULT_FACTORY,
+			failureManager);
+
+		coord.startCheckpointScheduler();
+		// Periodic
+		final CompletableFuture<CompletedCheckpoint> onCompletionPromise1 = coord.triggerCheckpoint(
+			System.currentTimeMillis(),
+			CheckpointProperties.forCheckpoint(CheckpointRetentionPolicy.NEVER_RETAIN_AFTER_TERMINATION),
+			null,
+			true,
+			false);
+		assertTrue(coord.isTriggering());
+		assertEquals(0, coord.getTriggerRequestQueue().size());
+
+		manuallyTriggeredScheduledExecutor.triggerAll();
+		try {
+			onCompletionPromise1.get();
+			fail("Should not reach here");
```

Comment:
What is causing a failure in this test case? I think this might be missing some comment in the lines above, what is causing the failure

Reply:
Hmmm, it's a bit hard to understand.
It's due to `TestingCheckpointIDCounter`. I should give it a better name.